Guide (#424)
FredTheDino/sylt-lang: docs/guide.adoc

Types can also function as values which is described in *REF*.

=== Variables and constant variables

Values can be assigned to variables, which works just the way you expect it to.

Variables can be captured, which is described in *REF*.

Variables can optionally be declared as _constant_. Constant variables will
always point to the same value, but the value itself can still be modified (if
it is a modifiable value, see below).

[source, sylt]
----
pp := 1     // a variable declaration
PI :: 3.14  // a constant declaration
----

//TODO this case has an actual name
Constant variables should be written in `CAPITAL_CASE` (apart from functions).

The type of the variable is inferred but can be supplied or even forced, which
is described in *REF*.

=== Basic values

These are best described with some code. Basic values are immutable.

[source, sylt]
----
// Strings
"string"
"contains spaces"
"sometimes \"escaped quotes\" as well"

// Integers
1
-2

// Floats
1.0
-5.5
1.  // trailing 0 optional
.5  // leading 0 optional

// Bools
true
false
----

=== Composite values

== If-statements and if-expressions

These are your normal control flow statements.

> .Normal if-statements

Reviewer: I would change "normal" to "standard" or "regular".

[source, sylt]
----
if a == 0 do
    print("zero")
end
----

There is also if-expressions, which evaluate depending on a condition.

.If-expressions as seen in e.g. Python
[source, sylt]
----
a := 5 if b == 0 else 1
----

A shortened form is available if the pass-branch is also the left hand side of
the condition.

.A motivating example for the short if-expression
[source, sylt]
----
// These two are equivalent
a := (b + 5.) / 10. if (b + 5.) / 10. > 1. else 0.
//   ^^^^^^^^^^^^^^    ^^^^^^^^^^^^^^
a := (b + 5.) / 10. if > 1. else 0.
----

== Operators

> The basic operators are all here. Normal mathematical grouping is applied.

Reviewer: ```suggestion
The basic operators are all here. Standard mathematical evaluation order applies.
```
Or something simiar. This line might need a linebreak.

.Arithmetic operators
[source, sylt]
----
1 + 1
2 - 2
3 * 3
10 / 2
-4
----

.Comparison operators
[source, sylt]
----
1 < 2
2 > 1
1 <= 1
1 >= 1
"a" == "a"

// Assert equal. Execution is halted if the values compare non-equal
[1, 2] <=> [1, 2]
----

.Parenthesis for grouping
[source, sylt]
----
(1 + 1) * 2
----

.Boolean operators
[source, sylt]
----
true or false
true and true
not true
----

== Imports

Code can be written in multiple files, to your liking. You don't have to

> consider include ordering and dependency cycles. Write your code anywhere you

Reviewer: ```suggestion
consider include ordering or dependency cycles. Write your code anywhere you
```

== Getting started

== Values and variables

There are 4 basic types which can be composed in lists, tuples, sets, dicts
and blobs. Functions, which functions as values, are described in *REF*.
Types can also function as values which is described in *REF*.

=== Variables and constant variables

Values can be assigned to variables, which works just the way you expect it to.

Variables can be captured, which is described in *REF*.

Variables can optionally be declared as _constant_. Constant variables will
always point to the same value, but the value itself can still be modified (if
it is a modifiable value, see below).

[source, sylt]
----
pp := 1     // a variable declaration
PI :: 3.14  // a constant declaration
----

//TODO this case has an actual name
Constant variables should be written in `CAPITAL_CASE` (apart from functions).

The type of the variable is inferred but can be supplied or even forced, which
is described in *REF*.

=== Basic values

These are best described with some code. Basic values are immutable.

[source, sylt]
----
// Strings
"string"
"contains spaces"
"sometimes \"escaped quotes\" as well"

// Integers
1
-2

// Floats
1.0
-5.5

> 1.    // trailing 0 optional

Reviewer 2: .1 aswell?

you know it's what you need. :)

Files are included relative to the current file. With a leading "/" the path is
started from the working directory of the game process, which might be useful
if a file is located a few directories down.

[source, sylt]
----
// a/b/c.sy
use d.sy   // imports "a/b/d.sy"
use /d.sy  // imports "d.sy"
----

A directory can also be used if it is supplied with a trailing "/", which will
import the file "exports.sy" in that directory. This can be used to create
modules containing related code and a central "exporting-file".

[source, sylt]
----
// a.sy
use d/  // imports "d/exports.sy"
----

Includes (both files and directories) can be aliased to other names.

[source, sylt]
----
// a.sy
use b as c   // imports "b.sy" under the namespace c
// use c/    // invalid since the namespace c exists
use c/ as d  // imports "c/exports.sy" under the namespace d
----

[source, sylt]
----
----

== Loops

There are no built-in loop constructs. Rather, the standard library supplies a
couple of higher-order functions. Still, they are written almost like a the
normal loops

=== for_each

[source, sylt]
----
l := [1, 2, 3]

> l -> for_each' fn a: int do print a end

Reviewer: ```suggestion
l -> for_each' fn a: int do print(a) end
```

l -> for_each' fn a: int do
    // many statements
end
----

.Functions don't have to be constructed in-place.
[source, sylt]
----
Player :: blob {
    pos: (int, int)
    vel: (int, int)
}

update_player :: fn p: Player do

>     pos += vel

Reviewer: ```suggestion
    p.pos += p.vel
```
